Modified mailroom.py and added functionality to send letters to everyone

shibinmathew opened the pull request.

natasha-aleksandrova commented on students/shibin_mathew/lesson4/trigram.py:
```diff
@@ -0,0 +1,34 @@
+    list_story = []
+    f = open(filename, 'r')
+    for line in f:
+            list_story.append(line.split())
```
this line is over-indented (should be 4 spaces)

natasha-aleksandrova commented on students/shibin_mathew/lesson4/trigram.py:
```diff
@@ -0,0 +1,34 @@
+    # with open(path+'/'+filename, 'r+') as f:
+    #     data = f.read()
+    list_story = []
+    f = open(filename, 'r')
```
prefer to use `with` statement when working with files

natasha-aleksandrova commented on students/shibin_mathew/lesson4/trigram.py:
```diff
@@ -0,0 +1,34 @@
+    list_punctutation = [',', '.', ';', "'", '"']
+    print(list_story)
+    for i in range(0, len(list_story)):
+        if i + 1 <= len(list_story) and i + 2 <= len(list_story):
```
since you're getting length 3 times, might be good to extract into a var

natasha-aleksandrova replied:
logically you only need to check `i + 2 <= len(list_story)` since if +2 will be less than size then + 1 will also be less than size

natasha-aleksandrova commented on students/shibin_mathew/lesson4/trigram.py:
```diff
@@ -0,0 +1,34 @@
+                # print(list_story[i] + ' ' + list_story[i+1])
+                if list_story[i] + ' ' + list_story[i + 1] == word:
+                    # if list_story[i + 2] not in list_punctutation:
+                        dict_word[word].append(list_story[i + 2])
```
this line is over-indented

natasha-aleksandrova commented on students/shibin_mathew/lesson4/trigram.py:
```diff
@@ -0,0 +1,34 @@
+    print('Original List')
+    print(dict_wordlist)
+    print('New List')
+    print(word_library(list_story, dict_wordlist))
```
I'm a little confused here 🙂 

why is `dict_wordlist` created manually here? you should read in a file and create an equivalent of `dict_wordlist` but using data from the story file. Then re-create story by choosing random starting point of word pair and then random selection of words from a list that matches your word pair key.

shibinmathew commented:
Hi,

I started working on the tiagram but have not completed it. This request
was for the first part of assignment 4 - sending letters to everyone.

Thanks,

Shibin

On Mon, Apr 30, 2018, 8:26 PM Natasha <notifications@github.com> wrote:

> *@natasha-aleksandrova* approved this pull request.
>
> nice work here, take a look at my comments on trigrams please.
>
> I may be misunderstanding but appears to miss some requirements.
> ------------------------------
>
> In students/shibin_mathew/lesson4/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185022173>
> :
>
> > +
> +def option_three():
> +    #  send letters to everyone
> +    for names in donors:
> +        with open(names['name']+'.txt', 'w') as text_file:
> +            text_file.write('Dear {},\n\n'
> +                            'Thank you for your kind donation of ${}.\n'
> +                            'It will be put to very good use\n\n'
> +                            '\t\t\tSincerely,\n'
> +                            '\t\t\t\t-The Team'.format(names['name'], round(sum(names['donation']), 2)))
> +    print("Letters have been sent to everyone!")
> +    return None
> +
> +
> +def switch_menu(option):
> +    dict_switch = {'1': option_one, '2': option_two, '3': option_three, '4': quit_menu}
>
> nice job using dict as a switch here, recommendation for readability of
> code: naming the function send_thank_you instead of option_one would tell
> reader right away what to expect from function.
> ------------------------------
>
> In students/shibin_mathew/lesson4/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185025796>
> :
>
> > +    program_quit = False
> +    while not program_quit:
> +        user_response = input("Please choose an action:\n\n"
> +                              "1. Send a Thank You\n"
> +                              "2. Create a Report\n"
> +                              "3. Send letters to everyone\n"
> +                              "4: Quit\n")
> +        switch_menu(user_response)()
> +
> +
> +if __name__ == "__main__":
> +    donors = [{'name': 'Shibin', 'donation': [25.25, 456.21]},
> +              {'name': 'Kimberly', 'donation': [125.50, 5.55, 14.5]},
> +              {'name': 'Andreck',
> +               'donation': [1250, 50, 1555]}, {'name': 'Jordy', 'donation':
> +              [55, 25, 4.50]}, {'name': 'Nelly', 'donation': [254.2, 500, 125]}]
>
> since we are introducing a dictionary in this lesson, you should try using
> dictionary for data structure instead of list which should make the code
> much simpler (hint: name is the natural key in this use case)
> ------------------------------
>
> In students/shibin_mathew/lesson4/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185026248>
> :
>
> > @@ -0,0 +1,101 @@
> +#!/usr/bin/env python3
> +import sys
> +
> +
> +def option_one():
>
> this function has too many logic levels, consider refactor to simplify.
> something like:
>
> while True:
>
>   name_response = input("full name or list or quit")
>
>   if name_response == 'list':
>     print_donor_names()
>   elif name_response == 'quit':  # exit function
>     return
>   else:
>     break  # stop asking, we got the name for the donor
>
>
> amt_response = input("donation amount or quit")
> if amt_response == 'quit':  # exit function
>   return
> else:
>   update_donors(name_response, amt_response)  # separate method
>   # or if you used default dict, very simple:
>   donors[name_response].append(amt_response)
>
>
>
>
> What I mean by logic levels is a lot of if/else blocks are deeply nested.
> This usually indicates a need for some refactor like combining logic flows
> or extracting some code into a function. That is done to make the code
> easier to read and understand.
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185144631>
> :
>
> > @@ -0,0 +1,34 @@
> +
> +def read_file(filename):
> +    # with open(path+'/'+filename, 'r+') as f:
> +    #     data = f.read()
> +    list_story = []
> +    f = open(filename, 'r')
> +    for line in f:
> +            list_story.append(line.split())
>
> this line is over-indented (should be 4 spaces)
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185144650>
> :
>
> > @@ -0,0 +1,34 @@
> +
> +def read_file(filename):
> +    # with open(path+'/'+filename, 'r+') as f:
> +    #     data = f.read()
> +    list_story = []
> +    f = open(filename, 'r')
>
> prefer to use with statement when working with files
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185144753>
> :
>
> > +def read_file(filename):
> +    # with open(path+'/'+filename, 'r+') as f:
> +    #     data = f.read()
> +    list_story = []
> +    f = open(filename, 'r')
> +    for line in f:
> +            list_story.append(line.split())
> +    return list_story
> +
> +
> +def word_library(list_story, dict_word):
> +    #  Adds new words to current list from the story
> +    list_punctutation = [',', '.', ';', "'", '"']
> +    print(list_story)
> +    for i in range(0, len(list_story)):
> +        if i + 1 <= len(list_story) and i + 2 <= len(list_story):
>
> since you're getting length 3 times, might be good to extract into a var
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185144908>
> :
>
> > +def read_file(filename):
> +    # with open(path+'/'+filename, 'r+') as f:
> +    #     data = f.read()
> +    list_story = []
> +    f = open(filename, 'r')
> +    for line in f:
> +            list_story.append(line.split())
> +    return list_story
> +
> +
> +def word_library(list_story, dict_word):
> +    #  Adds new words to current list from the story
> +    list_punctutation = [',', '.', ';', "'", '"']
> +    print(list_story)
> +    for i in range(0, len(list_story)):
> +        if i + 1 <= len(list_story) and i + 2 <= len(list_story):
>
> logically you only need to check i + 2 <= len(list_story) since if +2
> will be less than size then + 1 will also be less than size
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185144958>
> :
>
> > +    for line in f:
> +            list_story.append(line.split())
> +    return list_story
> +
> +
> +def word_library(list_story, dict_word):
> +    #  Adds new words to current list from the story
> +    list_punctutation = [',', '.', ';', "'", '"']
> +    print(list_story)
> +    for i in range(0, len(list_story)):
> +        if i + 1 <= len(list_story) and i + 2 <= len(list_story):
> +            for word in dict_word:
> +                # print(list_story[i] + ' ' + list_story[i+1])
> +                if list_story[i] + ' ' + list_story[i + 1] == word:
> +                    # if list_story[i + 2] not in list_punctutation:
> +                        dict_word[word].append(list_story[i + 2])
>
> this line is over-indented
> ------------------------------
>
> In students/shibin_mathew/lesson4/trigram.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#discussion_r185145456>
> :
>
> > +                # print(list_story[i] + ' ' + list_story[i+1])
> +                if list_story[i] + ' ' + list_story[i + 1] == word:
> +                    # if list_story[i + 2] not in list_punctutation:
> +                        dict_word[word].append(list_story[i + 2])
> +    return dict_word
> +
> +
> +if __name__ == '__main__':
> +    filename = 'sherlock_small.txt'
> +    dict_wordlist = {'I wish': ['I', 'I'], 'wish I': ['may', 'might'], 'may I': ['wish'], 'I may': ['I'],
> +                     'it was': ['at'], 'with the': ['cat']}
> +    list_story = read_file(filename)
> +    print('Original List')
> +    print(dict_wordlist)
> +    print('New List')
> +    print(word_library(list_story, dict_wordlist))
>
> I'm a little confused here 🙂
>
> why is dict_wordlist created manually here? you should read in a file and
> create an equivalent of dict_wordlist but using data from the story file.
> Then re-create story by choosing random starting point of word pair and
> then random selection of words from a list that matches your word pair key.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online/pull/491#pullrequestreview-116330330>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOhKiRxPvYBwQrFwP9ggsLJhtSy8Fr-Qks5tt6u0gaJpZM4TsA-_>
> .
>

natasha-aleksandrova commented:
I see that makes more sense. Please let us know in the next PR that part of trigram was merged earlier.
Thanks.

shibinmathew commented on students/shibin_mathew/lesson4/mailroom.py:
```diff
@@ -0,0 +1,101 @@
+import sys
+
+
+def option_one():
```
Thanks for the tip. This helped clean up my code!
